include/xsimd/arch/xsimd_avx2.hpp

```diff
@@ -313,44 +313,62 @@ namespace xsimd
         }
 
         // gather
-        template <class A, class T,
-                  typename std::enable_if<std::is_same<uint32_t, T>::value || std::is_same<int32_t, T>::value,
-                                          void>::type>
-        inline batch<T, A> gather(T const* src, batch<int32_t, A> const& index,
+        template <class T, class A, class U, detail::enable_sized_integral_t<T, 4> = 0, detail::enable_sized_integral_t<U, 4> = 0>
+        inline batch<T, A> gather(batch<T, A> const&, T const* src, batch<U, A> const& index,
                                   kernel::requires_arch<avx2>) noexcept
         {
             // scatter for this one is AVX512F+AVX512VL
-            return _mm256_i32gather_epi32(src, index, sizeof(T));
+            return _mm256_i32gather_epi32(reinterpret_cast<const int*>(src), index, sizeof(T));
         }
 
-        template <class A, class T,
-                  typename std::enable_if<std::is_same<uint64_t, T>::value || std::is_same<int64_t, T>::value,
-                                          void>::type>
-        inline batch<T, A> gather(T const* src, batch<int64_t, A> const& index,
+        template <class T, class A, class U, detail::enable_sized_integral_t<T, 8> = 0, detail::enable_sized_integral_t<U, 8> = 0>
+        inline batch<T, A> gather(batch<T, A> const&, T const* src, batch<U, A> const& index,
                                   kernel::requires_arch<avx2>) noexcept
         {
             // scatter for this one is AVX512F+AVX512VL
-            return _mm256_i64gather_epi64(src, index, sizeof(T));
+            return _mm256_i64gather_epi64(reinterpret_cast<const long long int*>(src), index, sizeof(T));
         }
 
-        template <class A>
-        inline batch<float, A> gather(float const* src,
-                                      batch<int32_t, A> const& index,
+        template <class A, class U,
+                  detail::enable_sized_integral_t<U, 4> = 0>
+        inline batch<float, A> gather(batch<float, A> const&, float const* src,
+                                      batch<U, A> const& index,
```

Comment:
Hi! This actually fixes a bug in the latest release. Before this change, this overload of `xsimd::kernel::gather` had only 1 template parameter. And since `xsimd::batch::gather` calls `kernel::gather<A, T>(src, index, A {});` (2 template arguments), this overload is always a mismatch and thus never selected. In my case, this resulted in `xsimd::kernel::gather` always falling back to the generic implementation on AVX2+FMA3.

I was about to send a bugfix, but then noticed that this issue is fixed on master with this PR. Thank you!

Reply:
Apologies there! I noticed that bug only when disassembling Krita a few months ago.

I think this could be sped up once a patch release is cut? @serge-sans-paille ?

Reply:
No need to apologies, you are doing great work! :)